include/lbann/data_readers/data_reader_HDF5_impl.hpp

```diff
@@ -0,0 +1,140 @@
+//////////////////////////////////////////////////////////////////////////////
+//
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+//
+////////////////////////////////////////////////////////////////////////////////
+#ifndef __REVISED_LBANN_DATA_READER_HDF5_IMPL_HPP__
+#define __REVISED_LBANN_DATA_READER_HDF5_IMPL_HPP__
+
+#include "lbann/data_readers/data_reader_HDF5.hpp"
+
+namespace lbann {
+
+template<typename T_from, typename T_to>
+void hdf5_data_reader::coerceme(const T_from* data_in, size_t n_bytes, std::vector<T_to> & data_out) {
+  size_t n_elts = n_bytes / sizeof(T_from);
+  data_out.resize(0);
+  data_out.reserve(n_elts);
+  for (size_t j=0; j<n_elts; j++) {
+    data_out.push_back(*data_in++);
```

Comment:
Do we need an explicit cast between types here?

Reply:
the value that is copied (data_in) should always be a primitive C data type, so any compiler should correctly cast. But @benson31  would better answer.

Reply:
A compiler _might_ complain about a narrowing cast (`int->size_t`, for example), but it's not a big deal.